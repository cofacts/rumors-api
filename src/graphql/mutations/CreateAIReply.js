import { GraphQLString, GraphQLNonNull } from 'graphql';

import openai from 'util/openai';
import { assertUser } from 'util/user';
import { AIReply } from 'graphql/models/AIResponse';
import { getAIResponse, createAIResponse } from 'graphql/util';

const monthFormatter = Intl.DateTimeFormat('zh-TW', {
  year: 'numeric',
  month: 'long',
  timeZone: 'Asia/Taipei',
});

/**
 * Create an new AIReply, initially in LOADING state, then becomes ERROR or SUCCESS,
 * and returns the AI reply.
 * If there is no enough content for AI, it resolves to null.
 */
export async function createNewAIReply({
  article,
  user,
  completionOptions = {},
}) {
  // article.hyperlinks deduped by URL.
  const dedupedHyperlinks = Object.values(
    (article.hyperlinks ?? []).reduce((map, hyperlink) => {
      if (
        !map[hyperlink.url] ||
        /* hyperlink exists, but fetch failed */ !map[hyperlink.url].title
      ) {
        map[hyperlink.url] = hyperlink;
      }
      return map;
    }, {})
  );

  /**
   * Determine if article has no content by replacing all URLs with its scrapped content.
   * This will become empty string if and only if:
   * - The article only contains URLs, no other text, and
   * - All URL scrapping results fail (no title, no summary)
   *
   * Abort AI reply generation in this case.
   */
  const replacedArticleText = dedupedHyperlinks
    .reduce(
      (text, { url, title, summary }) =>
        text.replaceAll(url, `${title} ${summary}`),
      article.text
    )
    .trim();

  if (replacedArticleText.length === 0) return null;

  // Argumenting hyperlinks with summary and titles
  const argumentedArticleText = dedupedHyperlinks.reduce(
    (text, { url, title, summary }) =>
      title
        ? text.replaceAll(url, `[${title} ${summary}](${url})`)
        : /* Fetch failed, don't replace */ text,
    article.text
  );

  const thisMonthParts = monthFormatter.formatToParts(new Date());
  const thisYearStr = thisMonthParts.find((p) => p.type === 'year').value;
  const thisROCYearStr = (+thisYearStr - 1911).toString();
  const thisMonthStr = thisMonthParts.find((p) => p.type === 'month').value;
  const createdMonth = monthFormatter.format(new Date(article.createdAt));

  const completionRequest = {
    model: 'gpt-3.5-turbo',
    messages: [
      {
        role: 'system',
        content: `現在是${thisYearStr}年（民國${thisROCYearStr}年）${thisMonthStr}月。你是協助讀者進行媒體識讀的小幫手。你說話時總是使用台灣繁體中文。有讀者傳了一則網路訊息給你。這則訊息${createdMonth}就在網路上流傳。`,
      },
      {
        role: 'user',
        content: argumentedArticleText,
      },
      {
        role: 'user',
        content:
          '請問作為閱聽人，我應該注意這則訊息的哪些地方呢？請節錄訊息中需要特別留意或懷疑的地方，說明為何閱聽人需要注意它。請只就以上內文回應，不要編造。謝謝',
      },
    ],
    user: user.id,
    temperature: 0,
    ...completionOptions,
  };

  const updateAIResponse = createAIResponse({
    user,
    docId: article.id,
    type: 'AI_REPLY',
    request: JSON.stringify(completionRequest),
  });

  // Resolves to completed or errored AI response.
  const apiResult = await openai
    .createChatCompletion(completionRequest)
    .then(({ data }) => data)
    .catch((error) => {
      console.error(error);

      /* Resolve with Error instance, which will be used to update AI response below */
      /* istanbul ignore else */
      if (error instanceof Error) return error;
      return new Error(error);
    });

  return updateAIResponse(
    apiResult instanceof Error
      ? {
          status: 'ERROR',
          text: apiResult.toString(),
        }
      : {
          status: 'SUCCESS',
          text: apiResult.choices[0].message.content,
          ...(apiResult.usage
            ? {
                usage: {
                  promptTokens: apiResult.usage.prompt_tokens,
                  completionTokens: apiResult.usage.completion_tokens,
                  totalTokens: apiResult.usage.total_tokens,
                },
              }
            : undefined),
        }
  );
}

export default {
  type: AIReply,
  description:
    'Create an AI reply for a specific article. If existed, returns an existing one. If information in the article is not sufficient for AI, return null.',
  args: {
    articleId: { type: new GraphQLNonNull(GraphQLString) },
  },
  async resolve(rootValue, { articleId }, { loaders, user }) {
    assertUser(user);

    const article = await loaders.docLoader.load({
      index: 'articles',
      id: articleId,
    });

    if (!article) throw new Error(`Article ${articleId} does not exist.`);

    const existingAiReply = await getAIResponse({
      type: 'AI_REPLY',
      docId: articleId,
    });

    return existingAiReply
      ? existingAiReply
      : createNewAIReply({ article, user });
  },
};
