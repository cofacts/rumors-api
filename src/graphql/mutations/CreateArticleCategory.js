import {
  GraphQLString,
  GraphQLNonNull,
  GraphQLList,
  GraphQLFloat,
} from 'graphql';

import client from 'util/client';
import { assertUser, getContentDefaultStatus } from 'util/user';
import ArticleCategory from 'graphql/models/ArticleCategory';

/**
 *
 * @param {object} param
 * @param {object} param.articleId - The articleId to attach category to
 * @param {object} param.categoryId - The categoryId to attach to article
 * @param {string} param.user - The user adding this article-reply connection
 * @returns {ArticleCategory[]} The article categories after creation
 */
export async function createArticleCategory({
  articleId,
  categoryId,
  user,
  aiModel,
  aiConfidence,
}) {
  assertUser(user);
  if (!articleId || !categoryId) {
    throw new Error(
      'articleId and categoryId are mandatory when creating ArticleCategory.'
    );
  }

  const now = new Date().toISOString();
  const defaultStatus = getContentDefaultStatus(user);

  const articleCategory = {
    userId: user.id,
    appId: user.appId,
    aiModel,
    aiConfidence,
    positiveFeedbackCount: 0,
    negativeFeedbackCount: 0,
    categoryId,
    status: defaultStatus,
    createdAt: now,
    updatedAt: now,
  };

  const {
    body: {
      result: articleResult,
      get: { _source },
    },
  } = await client.update({
    index: 'articles',
    type: 'doc',
    id: articleId,
    body: {
      script: {
        /**
         * Check if the category is already connected in the article.
         * If connected with not default status, then set to default status.
         * If connected with NORMAL status, then do nothing.
         * Otherwise, do update.
         */
        source: `
          def found = ctx._source.articleCategories.stream()
            .filter(ar -> ar.get('categoryId').equals(params.articleCategory.get('categoryId')))
            .findFirst();

          if (found.isPresent()) {
            HashMap ar = found.get();
            if (ar.get('status').equals(params.defaultStatus)) {
              ctx.op = 'none';
            } else {
              ar.put('status', params.defaultStatus);
              ar.put('userId', params.articleCategory.get('userId'));
              ar.put('appId', params.articleCategory.get('appId'));
              ar.put('updatedAt', params.articleCategory.get('updatedAt'));
            }
          } else {
            ctx._source.articleCategories.add(params.articleCategory);
            ctx._source.normalArticleCategoryCount = ctx._source.articleCategories.stream().filter(
              ar -> ar.get('status').equals('NORMAL')
            ).count();
          }
        `,
        lang: 'painless',
        params: { articleCategory, defaultStatus },
      },
      _source: ['articleCategories.*'],
    },
  });

  if (articleResult === 'noop') {
    throw new Error(
      `Cannot add articleCategory ${JSON.stringify(
        articleCategory
      )} to article ${articleId}`
    );
  }

  return _source.articleCategories;
}

export default {
  type: new GraphQLList(ArticleCategory),
  description: 'Adds specified category to specified article.',
  args: {
    articleId: { type: new GraphQLNonNull(GraphQLString) },
    categoryId: { type: new GraphQLNonNull(GraphQLString) },
    aiModel: { type: GraphQLString },
    aiConfidence: { type: GraphQLFloat },
  },
  async resolve(
    rootValue,
    { articleId, categoryId, aiModel, aiConfidence },
    { user }
  ) {
    const articleCategories = await createArticleCategory({
      articleId,
      categoryId,
      user,
      aiModel,
      aiConfidence,
    });

    // When returning, insert articleId so that ArticleReply object type can resolve article.
    return articleCategories.map((ac) => ({ ...ac, articleId }));
  },
};
