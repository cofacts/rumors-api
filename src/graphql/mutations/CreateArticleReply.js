import { GraphQLString, GraphQLNonNull, GraphQLList } from 'graphql';

import client from 'util/client';
import { assertUser, getContentDefaultStatus } from 'util/user';
import ArticleReply from 'graphql/models/ArticleReply';

/**
 *
 * @param {object} param
 * @param {object} param.article - The article instance to attach reply to
 * @param {object} param.reply - The reply instance to attach to article
 * @param {object} user - The user adding this article-reply connection
 * @returns {[ArticleReply]} The article replies after creation
 */
export async function createArticleReply({ article, reply, user }) {
  assertUser(user);
  if (!article || !reply) {
    throw new Error(
      'articleId and replyId are mandatory when creating ArticleReplies.'
    );
  }

  const now = new Date().toISOString();

  const articleReply = {
    userId: user.id,
    appId: user.appId,
    positiveFeedbackCount: 0,
    negativeFeedbackCount: 0,
    replyId: reply.id,
    replyType: reply.type,
    status: getContentDefaultStatus(user),
    createdAt: now,
    updatedAt: now,
  };

  const {
    body: {
      result: articleResult,
      get: { _source },
    },
  } = await client.update({
    index: 'articles',
    type: 'doc',
    id: article.id,
    body: {
      script: {
        /**
         * Check if the reply is already connected in the article.
         * If so, do nothing;
         * otherwise, do update.
         */
        source: `
          if(
            ctx._source.articleReplies.stream().anyMatch(
              ar -> ar.get('replyId').equals(params.articleReply.get('replyId'))
            )
          ) {
            ctx.op = 'none';
          } else {
            ctx._source.articleReplies.add(params.articleReply);
            ctx._source.normalArticleReplyCount = ctx._source.articleReplies.stream().filter(
              ar -> ar.get('status').equals('NORMAL')
            ).count();
          }
        `,
        lang: 'painless',
        params: { articleReply },
      },
      _source: ['articleReplies.*'],
    },
  });

  if (articleResult === 'noop') {
    throw new Error(
      `Cannot add articleReply ${JSON.stringify(articleReply)} to article ${
        article.id
      }`
    );
  }

  return _source.articleReplies;
}

export default {
  type: new GraphQLList(ArticleReply),
  description: 'Connects specified reply and specified article.',
  args: {
    articleId: { type: new GraphQLNonNull(GraphQLString) },
    replyId: { type: new GraphQLNonNull(GraphQLString) },
  },
  async resolve(rootValue, { articleId, replyId }, { user, loaders }) {
    const [article, reply] = await loaders.docLoader.loadMany([
      { index: 'articles', id: articleId },
      { index: 'replies', id: replyId },
    ]);

    const articleReplies = await createArticleReply({
      article,
      reply,
      user,
    });

    // When returning, insert articleId so that ArticleReply object type can resolve article.
    return articleReplies.map((ar) => ({ ...ar, articleId }));
  },
};
