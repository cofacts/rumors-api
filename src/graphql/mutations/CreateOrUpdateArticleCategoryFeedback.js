import { GraphQLString, GraphQLNonNull } from 'graphql';

import { assertUser, getContentDefaultStatus } from 'util/user';
import FeedbackVote from 'graphql/models/FeedbackVote';
import ArticleCategory from 'graphql/models/ArticleCategory';
import DataLoaders from 'graphql/dataLoaders';

import client from 'util/client';

export function getArticleCategoryFeedbackId({
  articleId,
  categoryId,
  userId,
  appId,
}) {
  return `${articleId}__${categoryId}__${userId}__${appId}`;
}

/**
 * @returns {ArticleCategory} The updated article category
 */
export async function createOrUpdateArticleCategoryFeedback({
  articleId,
  categoryId,
  vote,
  comment,
  user,

  // Use new instance of dataLoader if not provided
  loaders = new DataLoaders(),
}) {
  const now = new Date().toISOString();

  // (articleId, categoryId, userId, appId) should be unique
  // but user can update
  const id = getArticleCategoryFeedbackId({
    articleId,
    categoryId,
    userId: user.id,
    appId: user.appId,
  });

  await client.update({
    index: 'articlecategoryfeedbacks',
    type: 'doc',
    id,
    body: {
      doc: {
        score: vote,
        comment: comment,
        updatedAt: now,
      },
      upsert: {
        articleId,
        categoryId,
        userId: user.id,
        appId: user.appId,
        score: vote,
        createdAt: now,
        updatedAt: now,
        comment: comment,
        status: getContentDefaultStatus(user),
      },
    },
    refresh: 'true', // We are searching for articlecategoryfeedbacks immediately
  });

  const feedbacks = await loaders.articleCategoryFeedbacksLoader.load({
    articleId,
    categoryId,
  });

  const [positiveFeedbackCount, negativeFeedbackCount] = feedbacks
    .filter(({ status }) => status === 'NORMAL')
    .reduce(
      (agg, { score }) => {
        if (score === 1) {
          agg[0] += 1;
        } else if (score === -1) {
          agg[1] += 1;
        }
        return agg;
      },
      [0, 0]
    );

  const { body: articleCategoryUpdateResult } = await client.update({
    index: 'articles',
    type: 'doc',
    id: articleId,
    body: {
      script: {
        source: `
            int idx = 0;
            int categoryCount = ctx._source.articleCategories.size();
            for(; idx < categoryCount; idx += 1) {
              HashMap articleCategory = ctx._source.articleCategories.get(idx);
              if( articleCategory.get('categoryId').equals(params.categoryId) ) {
                break;
              }
            }

            if( idx === categoryCount ) {
              ctx.op = 'none';
            } else {
              ctx._source.articleCategories.get(idx).put(
                'positiveFeedbackCount', params.positiveFeedbackCount);
              ctx._source.articleCategories.get(idx).put(
                'negativeFeedbackCount', params.negativeFeedbackCount);
            }
          `,
        params: {
          categoryId,
          positiveFeedbackCount,
          negativeFeedbackCount,
        },
      },
    },
    _source: true,
  });

  /* istanbul ignore if */
  if (articleCategoryUpdateResult.result !== 'updated') {
    throw new Error(
      `Cannot article ${articleId}'s feedback count for feedback ID = ${id}`
    );
  }

  return articleCategoryUpdateResult.get._source.articleCategories.find(
    (articleCategory) => articleCategory.categoryId === categoryId
  );
}

export default {
  description: 'Create or update a feedback on an article-category connection',
  type: ArticleCategory,
  args: {
    articleId: { type: new GraphQLNonNull(GraphQLString) },
    categoryId: { type: new GraphQLNonNull(GraphQLString) },
    vote: { type: new GraphQLNonNull(FeedbackVote) },
    comment: { type: GraphQLString },
  },

  async resolve(
    rootValue,
    { articleId, categoryId, vote, comment },
    { user, loaders }
  ) {
    assertUser(user);

    const updatedArticleCategory = await createOrUpdateArticleCategoryFeedback({
      articleId,
      categoryId,
      vote,
      comment,
      user,
      loaders,
    });

    /* istanbul ignore if */
    if (!updatedArticleCategory) {
      throw new Error(
        `Cannot get updated article category with article ID = ${articleId} and category ID = ${categoryId}`
      );
    }

    return { articleId, ...updatedArticleCategory };
  },
};
