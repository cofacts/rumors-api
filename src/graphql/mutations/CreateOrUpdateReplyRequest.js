import { GraphQLString, GraphQLNonNull } from 'graphql';
import { assertUser, getContentDefaultStatus } from 'util/user';
import Article from '../models/Article';

import client, { processMeta } from 'util/client';

/**
 * @typedef {Object} ReplyRequestParam
 * @property {string} articleId - The article to add reply request to
 * @property {string} userId - The user that submits this request
 * @property {string} appId - The app that the user logged in to
 * @property {string} reason - The reason why the user want to submit this article
 */

/**
 * @param {ReplyRequestParam} params
 * @returns {string} the reply request ID for the givne params
 */
export function getReplyRequestId({ articleId, userId, appId }) {
  return `${articleId}__${userId}__${appId}`;
}

/**
 * @typedef {Object} CreateOrUpdateReplyRequestResult
 * @property {Object} article - The update article instance
 * @property {boolean} isCreated - If the reply request is created
 */

/**
 * Indexes a reply request and increments the replyRequestCount for article
 *
 * @param {ReplyRequestParam} param
 * @returns {CreateReplyRequstResult}
 */
export async function createOrUpdateReplyRequest({
  articleId,
  user,
  reason = '',
}) {
  assertUser(user);

  const now = new Date().toISOString();
  const id = getReplyRequestId({
    articleId,
    userId: user.id,
    appId: user.appId,
  });
  const updatedDoc = {
    updatedAt: now,
  };

  // Update reason if there is new one
  if (reason) {
    updatedDoc.reason = reason;
  }

  const replyRequestStatus = getContentDefaultStatus(user);

  const {
    body: { result },
  } = await client.update({
    index: 'replyrequests',
    type: 'doc',
    id,
    body: {
      doc: updatedDoc,
      upsert: {
        articleId,
        userId: user.id,
        appId: user.appId,
        reason,
        feedbacks: [],
        positiveFeedbackCount: 0,
        negativeFeedbackCount: 0,
        createdAt: now,
        updatedAt: now,
        status: replyRequestStatus,
      },
    },
    refresh: 'true',
  });

  const isCreated = result === 'created';

  // Update article for normal reply requests
  //
  const article = await (async () => {
    if (replyRequestStatus !== 'NORMAL') {
      return (
        await client.get({
          index: 'articles',
          type: 'doc',
          id: articleId,
        })
      ).body;
    }

    const { body: articleUpdateResult } = await client.update({
      index: 'articles',
      type: 'doc',
      id: articleId,
      body: {
        script: {
          source: `
            ${isCreated ? 'ctx._source.replyRequestCount += 1;' : ''}
            ctx._source.lastRequestedAt = params.now;
          `,
          params: { now },
        },
      },
      _source: true,
    });
    if (articleUpdateResult.result !== 'updated') {
      throw new Error(`Cannot update article ${articleId}`);
    }

    return articleUpdateResult.get;
  })();

  return {
    article: processMeta({ ...article, _id: articleId }),
    isCreated,
  };
}

export default {
  description: 'Create or update a reply request for the given article',
  type: Article,
  args: {
    articleId: { type: new GraphQLNonNull(GraphQLString) },
    reason: {
      type: GraphQLString,
      description: 'The reason why the user want to submit this article',
    },
  },
  async resolve(rootValue, { articleId, reason }, { user }) {
    const { article } = await createOrUpdateReplyRequest({
      articleId,
      user,
      reason,
    });
    return article;
  },
};
