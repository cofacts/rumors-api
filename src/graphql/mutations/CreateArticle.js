import { GraphQLString, GraphQLNonNull } from 'graphql';
import { h64 } from 'xxhashjs';

import { assertUser, getContentDefaultStatus } from 'util/user';
import client from 'util/client';
import scrapUrls from 'util/scrapUrls';

import { ArticleReferenceInput } from 'graphql/models/ArticleReference';
import MutationResult from 'graphql/models/MutationResult';
import { createOrUpdateReplyRequest } from './CreateOrUpdateReplyRequest';

/* Instantiate hash function */
const xxhash64 = h64();

/**
 * Generates ID from article text. Outputs identical ID when the given the same article text.
 *
 * Used for preventing identical articles sending in within a short amount of time,
 * i.e. while other articles are sending in.
 *
 * @param {string} text The article text
 * @returns {string} generated article ID
 */
export function getArticleId(text) {
  return xxhash64
    .update(text)
    .digest()
    .toString(36);
}

/**
 * Creates a new article in ElasticSearch,
 * or updates its reference by adding a new reference
 *
 * @param {object} param
 * @param {string} param.text
 * @param {ArticleReferenceInput} param.reference
 * @param {object} user - The user submitting this article
 * @returns {Promise<string>} the new article's ID
 */
async function createNewArticle({ text, reference: originalReference, user }) {
  const articleId = getArticleId(text);
  const now = new Date().toISOString();
  const reference = {
    ...originalReference,
    createdAt: now,
    userId: user.id,
    appId: user.appId,
  };

  await client.update({
    index: 'articles',
    type: 'doc',
    id: articleId,
    body: {
      script: {
        source: `
          ctx._source.updatedAt = params.updatedAt;
          ctx._source.references.add(params.reference);
        `,
        lang: 'painless',
        params: {
          updatedAt: now,
          reference,
        },
      },
      upsert: {
        text,
        createdAt: now,
        updatedAt: now,
        userId: user.id,
        appId: user.appId,
        references: [reference],
        articleReplies: [],
        articleCategories: [],
        normalArticleReplyCount: 0,
        normalArticleCategoryCount: 0,
        replyRequestCount: 0,
        tags: [],
        hyperlinks: [],
        articleType: 'TEXT',
        attachmentUrl: '',
        attachmentHash: '',
        status: getContentDefaultStatus(user),
        contributors: [],
      },
    },
    refresh: 'true', // Make sure the data is indexed when we create ReplyRequest
  });

  return articleId;
}

/**
 * @param {string} articleId
 * @param {ScrapResult[]} hyperlinks
 * @return {Promise | null} update result
 */
export function updateArticleHyperlinks(articleId, scrapResults) {
  if (!scrapResults || scrapResults.length === 0) return Promise.resolve(null);

  return client.update({
    index: 'articles',
    type: 'doc',
    id: articleId,
    body: {
      doc: {
        hyperlinks: scrapResults.map(
          ({ url, normalizedUrl, title, summary }) => ({
            url,
            normalizedUrl,
            title,
            summary,
          })
        ),
      },
    },
  });
}

export default {
  type: MutationResult,
  description: 'Create an article and/or a replyRequest',
  args: {
    text: { type: new GraphQLNonNull(GraphQLString) },
    reference: { type: new GraphQLNonNull(ArticleReferenceInput) },
    reason: {
      // FIXME: Change to required field after LINE bot is implemented
      // type: new GraphQLNonNull(GraphQLString),
      type: GraphQLString,
      description:
        'The reason why the user want to submit this article. Mandatory for 1st sender',
    },
  },
  resolve(rootValue, { text, reference, reason }, { user, loaders }) {
    assertUser(user);
    const newArticlePromise = createNewArticle({
      text,
      reference,
      user,
    });

    const scrapPromise = scrapUrls(text, {
      cacheLoader: loaders.urlLoader,
      client,
    });

    // Dependencies
    //
    // newArticlePromise* --> replyRequestPromise* -.
    //                  \                            \ (ensure no parallel update to article)
    // scrapPromise -----`----------------------------`--> hyperlinkPromise* --> done
    //
    // *: Updates article. Will trigger version_conflict_engine_exception if run in parallel.
    //

    const replyRequestPromise = newArticlePromise.then(articleId =>
      createOrUpdateReplyRequest({ articleId, user, reason })
    );

    const hyperlinkPromise = Promise.all([
      newArticlePromise,
      scrapPromise,
      replyRequestPromise, // Not used, just to avoid parallel update to article
    ]).then(([articleId, scrapResults]) => {
      return updateArticleHyperlinks(articleId, scrapResults);
    });

    // Wait for all promises
    return Promise.all([
      newArticlePromise, // for fetching articleId
      hyperlinkPromise,
      replyRequestPromise,
    ]).then(([id]) => ({ id }));
  },
};
