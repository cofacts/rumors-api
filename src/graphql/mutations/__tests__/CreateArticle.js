import gql from 'util/GraphQL';
import { loadFixtures, unloadFixtures } from 'util/fixtures';
import client from 'util/client';
import MockDate from 'mockdate';
import fixtures, { fixture1Text } from '../__fixtures__/CreateArticle';
import { getReplyRequestId } from '../CreateOrUpdateReplyRequest';
import { getArticleId } from 'graphql/mutations/CreateArticle';

describe('creation', () => {
  beforeEach(() => loadFixtures(fixtures));
  afterEach(() => unloadFixtures(fixtures));

  it('creates articles and a reply request and fills in URLs', async () => {
    MockDate.set(1485593157011);
    const userId = 'test';
    const appId = 'foo';

    const { data, errors } = await gql`
      mutation ($text: String!, $reference: ArticleReferenceInput!) {
        CreateArticle(
          text: $text
          reference: $reference
          reason: "気になります"
        ) {
          id
        }
      }
    `(
      {
        text: 'FOO FOO http://foo.com/article/1',
        reference: { type: 'LINE' },
      },
      { user: { id: userId, appId } }
    );
    MockDate.reset();

    expect(errors).toBeUndefined();

    const {
      body: { _source: article },
    } = await client.get({
      index: 'articles',
      type: 'doc',
      id: data.CreateArticle.id,
    });

    expect(article.replyRequestCount).toBe(1);
    expect(article).toMatchSnapshot();

    const replyRequestId = getReplyRequestId({
      articleId: data.CreateArticle.id,
      userId,
      appId,
    });

    const {
      body: { _source: replyRequest },
    } = await client.get({
      index: 'replyrequests',
      type: 'doc',
      id: replyRequestId,
    });

    expect(replyRequest).toMatchInlineSnapshot(`
      Object {
        "appId": "foo",
        "articleId": "y0iq3fr2pqgi",
        "createdAt": "2017-01-28T08:45:57.011Z",
        "feedbacks": Array [],
        "negativeFeedbackCount": 0,
        "positiveFeedbackCount": 0,
        "reason": "気になります",
        "status": "NORMAL",
        "updatedAt": "2017-01-28T08:45:57.011Z",
        "userId": "test",
      }
    `);

    // Cleanup
    await client.delete({
      index: 'articles',
      type: 'doc',
      id: data.CreateArticle.id,
    });

    await client.delete({
      index: 'replyrequests',
      type: 'doc',
      id: replyRequestId,
    });
  });

  it('avoids creating duplicated articles and adds replyRequests automatically', async () => {
    MockDate.set(1485593157011);
    const userId = 'test';
    const appId = 'foo';
    const articleId = getArticleId(fixture1Text);

    const { data, errors } = await gql`
      mutation ($text: String!, $reference: ArticleReferenceInput!) {
        CreateArticle(
          text: $text
          reference: $reference
          reason: "気になります"
        ) {
          id
        }
      }
    `(
      {
        text: fixture1Text,
        reference: { type: 'LINE' },
      },
      { user: { id: userId, appId } }
    );
    MockDate.reset();
    expect(errors).toBeUndefined();

    // Expects no new article is created,
    // and it returns the existing ID
    expect(data.CreateArticle.id).toBe(articleId);

    const {
      body: { _source: article },
    } = await client.get({
      index: 'articles',
      type: 'doc',
      id: articleId,
    });

    // Expects lastRequestedAt, references are updated
    expect(article).toMatchSnapshot();

    // Expects new replyRequest is generated
    const replyRequestId = getReplyRequestId({ articleId, appId, userId });
    const {
      body: { _source: replyRequest },
    } = await client.get({
      index: 'replyrequests',
      type: 'doc',
      id: replyRequestId,
    });

    expect(replyRequest).toMatchInlineSnapshot(`
      Object {
        "appId": "foo",
        "articleId": "20xz7y20qzyt6",
        "createdAt": "2017-01-28T08:45:57.011Z",
        "feedbacks": Array [],
        "negativeFeedbackCount": 0,
        "positiveFeedbackCount": 0,
        "reason": "気になります",
        "status": "NORMAL",
        "updatedAt": "2017-01-28T08:45:57.011Z",
        "userId": "test",
      }
    `);

    // Cleanup
    await client.delete({
      index: 'replyrequests',
      type: 'doc',
      id: replyRequestId,
    });
  });

  it('sets status to blocked when author is blocked', async () => {
    MockDate.set(1485593157011);

    const userId = 'iAmBlocked';
    const appId = 'foo';
    const { data } = await gql`
      mutation ($text: String!, $reference: ArticleReferenceInput!) {
        CreateArticle(text: $text, reference: $reference, reason: "") {
          id
        }
      }
    `(
      {
        text: 'This is a scam (and I am actually scam too)',
        reference: { type: 'LINE' },
      },
      {
        user: {
          id: userId,
          appId,
          blockedReason: 'Announcement url',
        },
      }
    );
    MockDate.reset();

    const {
      body: { _source: article },
    } = await client.get({
      index: 'articles',
      type: 'doc',
      id: data.CreateArticle.id,
    });

    expect(article).toHaveProperty('status', 'BLOCKED');

    const replyRequestId = getReplyRequestId({
      articleId: data.CreateArticle.id,
      userId,
      appId,
    });

    const {
      body: { _source: replyRequest },
    } = await client.get({
      index: 'replyrequests',
      type: 'doc',
      id: replyRequestId,
    });

    expect(replyRequest).toHaveProperty('status', 'BLOCKED');

    // Cleanup
    await client.delete({
      index: 'articles',
      type: 'doc',
      id: data.CreateArticle.id,
    });

    await client.delete({
      index: 'replyrequests',
      type: 'doc',
      id: replyRequestId,
    });
  });
});

const testId = async (userId, appId) => {
  MockDate.set(1485593157011);
  const { errors } = await gql`
    mutation ($text: String!, $reference: ArticleReferenceInput!) {
      CreateArticle(
        text: $text
        reference: $reference
        reason: "気になります"
      ) {
        id
      }
    }
  `(
    {
      text: 'FOO FOO',
      reference: { type: 'LINE' },
    },
    { user: { id: userId, appId } }
  );
  MockDate.reset();
  expect(errors).toMatchSnapshot();
};

it('fails with an invalid user id', () => testId('', ''));
it('fails with an invalid app id', () => testId('test', ''));
