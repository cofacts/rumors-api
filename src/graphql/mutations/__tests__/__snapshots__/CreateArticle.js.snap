// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creation avoids creating duplicated articles and adds replyRequests automatically 1`] = `
Object {
  "lastRequestedAt": "2017-01-28T08:45:57.011Z",
  "references": Array [
    Object {
      "type": "LINE",
    },
    Object {
      "appId": "foo",
      "createdAt": "2017-01-28T08:45:57.011Z",
      "type": "LINE",
      "userId": "test",
    },
  ],
  "replyRequestCount": 2,
  "text": "I think I am I exist",
  "updatedAt": "2017-01-28T08:45:57.011Z",
}
`;

exports[`creation creates articles and a reply request and fills in URLs 1`] = `
Object {
  "appId": "foo",
  "articleCategories": Array [],
  "articleReplies": Array [],
  "articleType": "TEXT",
  "attachmentHash": "",
  "attachmentUrl": "",
  "contributors": Array [],
  "createdAt": "2017-01-28T08:45:57.011Z",
  "hyperlinks": Array [
    Object {
      "normalizedUrl": "http://foo.com/article/1",
      "summary": "FOO article content",
      "title": "FOO title",
      "url": "http://foo.com/article/1",
    },
  ],
  "lastRequestedAt": "2017-01-28T08:45:57.011Z",
  "normalArticleCategoryCount": 0,
  "normalArticleReplyCount": 0,
  "references": Array [
    Object {
      "appId": "foo",
      "createdAt": "2017-01-28T08:45:57.011Z",
      "type": "LINE",
      "userId": "test",
    },
  ],
  "replyRequestCount": 1,
  "status": "NORMAL",
  "text": "FOO FOO http://foo.com/article/1",
  "updatedAt": "2017-01-28T08:45:57.011Z",
  "userId": "test",
}
`;

exports[`fails with an invalid app id 1`] = `
Array [
  [GraphQLError: userId is set, but x-app-id or x-app-secret is not set accordingly.],
]
`;

exports[`fails with an invalid user id 1`] = `
Array [
  [GraphQLError: userId is not set via query string.],
]
`;
